message_ix/macro.py

```diff
@@ -1,61 +1,494 @@
+            'unit': 'G$',
+            'data_key': 'total_cost',
+        },
+        'gdp_calibrate': {
+            'idx': ['node', 'year', ],
+            'unit': 'T$',
+        },
+        'historical_gdp': {
+            'idx': ['node', 'year', ],
+            'unit': 'T$',
+            'data_key': 'gdp0',
+        },
+        'MERtoPPP': {
+            'idx': ['node', 'year', ],
+            'unit': '-',
+        },
+        'kgdp': {
+            'idx': ['node', ],
+            'unit': '-',
+        },
+        'kpvs': {
+            'idx': ['node', ],
+            'unit': '-',
+        },
+        'depr': {
+            'idx': ['node', ],
+            'unit': '-',
+        },
+        'drate': {
+            'idx': ['node', ],
+            'unit': '-',
+        },
+        'esub': {
+            'idx': ['node', ],
+            'unit': '-',
+        },
+        'lotol': {
+            'idx': ['node', ],
+            'unit': '-',
+        },
+        'lakl': {
+            'idx': ['node', ],
+            'unit': '-',
+            'data_key': 'aconst',
+        },
+        'prfconst': {
+            'idx': ['node', 'sector', ],
+            'unit': '-',
+            'data_key': 'bconst',
+        },
+        'grow': {
+            'idx': ['node', 'year', ],
+            'unit': '-',
+            'data_key': 'growth',
+        },
+        'aeei': {
+            'idx': ['node', 'sector', 'year', ],
+            'unit': '-',
+        },
+    },
+    'vars': {
+        'DEMAND': ['node', 'commodity', 'level', 'year', 'time', ],
+        'PRICE': ['node', 'commodity', 'level', 'year', 'time', ],
+        'COST_NODAL': ['node', 'year', ],
+        'COST_NODAL_NET': ['node', 'year', ],
+        'GDP': ['node', 'year', ],
+        'I': ['node', 'year', ],
+        'C': ['node', 'year', ],
+        'K': ['node', 'year', ],
+        'KN': ['node', 'year', ],
+        'Y': ['node', 'year', ],
+        'YN': ['node', 'year', ],
+        'EC': ['node', 'year', ],
+        'UTILITY': None,
+        'PHYSENE': ['node', 'sector', 'year', ],
+        'PRODENE': ['node', 'sector', 'year', ],
+        'NEWENE': ['node', 'sector', 'year', ],
+        'grow_calibrate': ['node', 'year', ],
+        'aeei_calibrate': ['node', 'sector', 'year', ],
+    },
+    'equs': {
+        'COST_ACCOUNTING_NODAL': ['node', 'year', ],
+    },
 }
 
+MACRO_DATA_FOR_DERIVATION = {
+    'cost_ref': ['node', ],
+    'demand_ref': ['node', 'sector', ],
+    'price_ref': ['node', 'sector', ],
+}
+
+VERIFY_INPUT_DATA = [
+    'price_ref', 'lotol', 'esub', 'drate', 'depr', 'kpvs', 'kgdp',
+    'gdp_calibrate', 'aeei', 'cost_ref', 'demand_ref', 'MERtoPPP',
+]
+
+
+def _validate_data(name, df, nodes, sectors, years):
+    def validate(kind, values, df):
+        if kind not in df:
+            return
+
+        diff = set(values) - set(df[kind])
+        if diff:
+            raise ValueError(
+                'Not all {}s included in {} data: {}'.format(kind, name, diff)
```

Comment:
Might there not be cases where not all nodes are calibrated (e.g. in the global model the "virtual trade node" _GLB_  is not calibrated so far. Is this really an issue?